### PR TITLE
calculate padding for ConvTranspose as well when output_size is given

### PR DESCRIPTION
**Pull request by felixgwu**
Now, ConvTranspose modules would calculate both padding and output_padding when the output_size is given to the forward method.

BTW, there was a bug here in line 252. It should drop the first two dimensions instead of keeping the last two of them. I fixed it.

### CODE REVIEW

File: torch/nn/modules/conv.py [hunk 1]
```diff
@@ -344,31 +344,25 @@ def forward(self, input, output_size=None):
...
-                raise ValueError((
-                    "requested an output size of {}, but valid sizes range "
-                    "from {} to {} (for an input of {})").format(
-                        output_size, min_sizes, max_sizes, input.size()[2:]))
-
-        return tuple([output_size[d] - min_sizes[d] for d in range(k)])
+        # 2*padding - output_padding = (input - 1)*stride - output + kernel_size
+        #                            = original_size
+        # padding = math.ceil(original_size / 2)
+        def original_size(d):
```
Comment (apaszke) [inline torch/nn/modules/conv.py]: I'd call it `no_pad_diff`. It's not an output size.

**Review by apaszke — COMMENTED** [review]
As far as I understand, this adds support for making the output smaller than it would originally be, right? What's the use case for that?

File: torch/nn/modules/conv.py (hunk 1 above)
Comment (felixgwu) [inline torch/nn/modules/conv.py]: Yes, it allows smaller output. When padding is less than zero, it can be considered as cropping instead of padding. Shall I add a comment regarding it in the code?
Yeah, `no_pad_diff` sounds like a better name of it.

File: torch/nn/modules/conv.py [hunk 2]
```diff
@@ -344,31 +344,25 @@ def forward(self, input, output_size=None):
...
-                    "from {} to {} (for an input of {})").format(
-                        output_size, min_sizes, max_sizes, input.size()[2:]))
-
-        return tuple([output_size[d] - min_sizes[d] for d in range(k)])
+        # 2*padding - output_padding = (input - 1)*stride - output + kernel_size
+        #                            = no_pad_diff
+        # padding = math.ceil(no_pad_diff / 2)
+        def no_pad_diff(d):
+            return ((input.size(d + 2) - 1) * self.stride[d] - output_size[d] + self.kernel_size[d])
+        padding = tuple([(no_pad_diff(d) + 1) // 2 for d in range(k)])
```
Comment (apaszke) [inline torch/nn/modules/conv.py]: My only concern is that I think that it's not compatible with what we've had before, is it? For example take a look at this code:
```
m = nn.ConvTranspose2d(3, 4, 3, 3)
i = Variable(torch.randn(2, 3, 6, 6))
o = m(i)
```
`o` will be of size `(2, 4, 18, 18)`, but we might actually want to get `(2, 4, 20, 20)`. The old formula would simply add `2` of output padding, but as far as I understand, this one will use -1 for padding and 0 for output padding (`no_pad_diff` will be -2 for each dim)?

Comment (felixgwu) [inline torch/nn/modules/conv.py]: That's right. It's not compatible to the old version.
The original version tries to keep the self.padding and pad the output on one side. 
However, this version pads the output symmetrically on both sides.
You may just modify your code as follows and print the output and look at the values on the boundaries.
```python
m = nn.ConvTranspose2d(3, 4, 3, 3)
i = Variable(torch.randn(2, 3, 6, 6))
o = m(i, output_size=(20, 20))
print(o)
```

I wonder what would be a better practice.
From my point of view, ConvTranspose2d is wildly used in GANs or AutoEncoders, so padding outputs symmetrically when generating images is more reasonable.

If you think compatibility is more important, then there two ways to adjust this:
1. Try to use self.padding first. If it is impossible to find such output_padding that matches the output_size, then we use the old method. Otherwise, we use the new one.
2. Find a new padding that is the closest to self.padding such that there exists a output_padding (between 0 and self.stride-1) matches the output_size.

Comment (soumith) [inline torch/nn/modules/conv.py]: Yes, I think (1) followed by (2) seems better. We NEVER want to break backward compatibility unless absolute essential.

**Comment by felixgwu** [comment]
Now, it would make minimal change to the padding when matching the output_size.

**Comment by felixgwu** [comment]
Hi @apaszke, I am wondering what the status of this PR is. Thanks.

**Comment by soumith** [comment]
@felixgwu at this point both @apaszke and I went through the PR a couple of times, but we are very confused as to what it does. That is why the delay in reviewing.

What do you mean by: "want to minimize padding_shift"?

Also, can you please give an example.

For example, we have an input of 11x11, with kernel of 3x3, okay then what happens with padding/output_padding ?

Also, there is now a merge conflict to be resolved.

**Comment by hjmjohnson** [comment]
Is this still relevant? Can it be closed?
